### ISSUE TITLE
Booting stuck in infinite loop when FW is in SPI flash

### ISSUE THREAD

**Issue by danielkucera** [zeldin/fx3lafw#4]
Hello,

how do I load the firmware?
These are my  built files:
```
-rw-rw-r--   1 danman danman 108K jan 28 18:11 fx3lafw-cypress-fx3.fw
-rwxrwxr-x   1 danman danman 356K jan 28 18:11 fx3lafw.elf
-rw-rw-r--   1 danman danman  89K jan 28 18:11 fx3lafw.map
```
I tried flashing `fx3lafw-cypress-fx3.fw` directly to SPI flash but it doesn't boot.

**Comment by zeldin** [comment]
The firmware is designed to be uploaded into RAM.  If you use [my libsigrok branch](https://github.com/zeldin/libsigrok/commits/fx3lafw) then sigrok will automatically upload the `fw` file when it detects a FX3 SuperSpeed Explorer board.  Note that due to the device switching from a USB2 to a USB3 bus after fw upload, sigrok will fail to detect the device coming back.  Just start sigrok again and it should detect the already programmed device.

**Comment by zeldin** [comment]
You can also use [fx3load](https://sourceforge.net/projects/fx3load/) to load the fw into RAM manually if you like.

**Comment by danielkucera** [comment]
Is there a way to store it permanently in SPI flash?

**Comment by danielkucera** [comment]
I've rebased your branch on latest master and disabled the firmware upload because it was not trivial to rebase the changes:
https://github.com/sigrokproject/libsigrok/compare/master...danielkucera:libsigrok:fx3

**Comment by danielkucera** [comment]
The firmware works when loaded by fx3load but doesn't enumerate when I flash the same file to 0x0 of SPI flash.
Do you have an idea what could be wrong?

**Comment by zeldin** [comment]
Nope, sorry.  I never tried to flash anything on the SuperSpeed Explorer.  Probably it needs to be in some other format?  Or if the flash is XIP the code may need to be relocated to a different address, or prefixed with a loader binary...

**Comment by danielkucera** [comment]
At the end, the format was right but the firmware was stuck on this loop:
https://github.com/zeldin/fx3lafw/blob/master/bsp/usb.c#L564
After changing while to if, it started to work.

**Comment by zeldin** [comment]
Interresting.  I guess you need to do things a little differently when not renumerating.
Changing the `while` to an `if` is equivalent to removing the code altogether (since executing the empty statement or not will not make any difference), so that is hardly the correct fix...

**Comment by danielkucera** [comment]
I was trying to understand the meaning of `FX3_UIB_POWER_ACTIVE` bit but haven't found much.
I searched github but found only: `CY_U3P_UIB_POWER_ADDRESS` used in `CyU3PDmaSocketIsValid` used in `CyU3PDmaChannelGetHandle` with almost no usages.

Do you have some reference where this is used?

**Comment by zeldin** [comment]
The register is documented on page 385 of [the TRM](http://www.cypress.com/?rID=80775) as a mechanism for performing a reset on the UIB block, and checking when its initialization is done.  I don't remember any more details than this (it's been almost 6 years...).

Since the reset operation works when booting from RAM but not from SPI, my guess would be that it is depending on some part of the chip being powered or initialized which isn't when booting from SPI, but which is (implicitly) when booting from RAM because the bootloader has already been using USB.  In which case powering/initializing that part before doing the UIB reset should fix the issue.

**Comment by danielkucera** [comment]
Now I noticed that the bit is acctually not writable (from software), check this table:
![image](https://github.com/zeldin/fx3lafw/assets/1734256/5c4487f0-f7a5-47b3-b414-128b112ee843)
If you check the register definition, there is only `R`:
![image](https://github.com/zeldin/fx3lafw/assets/1734256/2069bfdd-33ec-43db-ad13-3a476cfe1cac)
When comparing to `UIBIN_POWER`, it has `R/W` there:
![image](https://github.com/zeldin/fx3lafw/assets/1734256/5281fb26-e04f-4f6f-a305-bcf9d40eb998)
Or `I2S_POWER`:
![image](https://github.com/zeldin/fx3lafw/assets/1734256/6407278c-8090-4197-bd93-828a91e91857)

So it is probably not possible to reset it from software.
What do you think?

**Comment by zeldin** [comment]
Weird.  It also says that the `ACTIVE` bit is read-only for the HW, which makes absolutely no sense given the description.
Given that the actual description of the bits is identical to that of the other IP blocks, and that the section before the bit descriptions state that "Every IP block" should implement these same functions, I'd be inclined to believe that the differing direction indication here is just a misprint.  Unfortunately there are several of those in the TRM...  :confused:

**Comment by danielkucera** [comment]
Could you please test if setting/resetting the bit has any effect on the ACTIVE bit?

**Comment by zeldin** [comment]
Well, you'll probably get the results quicker if you do it yourself.  :smile:

**Comment by danielkucera** [comment]
interesting, the same code:
```
  char buf[64];
  snprintf(buf, sizeof(buf), "power val %lx\n", Fx3ReadReg32(FX3_UIB_POWER));
  Fx3UartTxString(buf);
  Fx3SetReg32(FX3_UIB_POWER, FX3_UIB_POWER_RESETN);
  Fx3UtilDelayUs(100);
  snprintf(buf, sizeof(buf), "power val %lx\n", Fx3ReadReg32(FX3_UIB_POWER));
  Fx3UartTxString(buf);
  Fx3ClearReg32(FX3_UIB_POWER, FX3_UIB_POWER_RESETN);
  Fx3UtilDelayUs(50);
  snprintf(buf, sizeof(buf), "power val %lx\n", Fx3ReadReg32(FX3_UIB_POWER));
  Fx3UartTxString(buf);
  Fx3SetReg32(FX3_UIB_POWER, FX3_UIB_POWER_RESETN);
  Fx3UtilDelayUs(100);
  snprintf(buf, sizeof(buf), "power val %lx\n", Fx3ReadReg32(FX3_UIB_POWER));
  Fx3UartTxString(buf);

```
behaves diferrently when run in boot from usb:
```
power val 80000001
power val 80000001
power val 0
power val 80000001
```
and when from SPI boot:

```
power val 0
power val 80000000
power val 0
power val 80000000
```

**Comment by danielkucera** [comment]
It is super-weird. Even if I put the value printing inside of main loop, it is still `80000000`, even during normal operation of the device.
```
req: c0 b2 value: 0000 index: 0000 length: 0001
CMD_GET_REVID_VERSION
Fx3UsbUsbCoreIsr
  PROT
    STATUS_STAGE
power val 80000000
power val 80000000
power val 80000000
power val 80000000
power val 80000000
power val 80000000
Fx3UsbUsbCoreIsr
  PROT
    SUTOK_EV
req: 40 b1 value: 0000 index: 0000 length: 0003
CMD_START
flags=80, sample_delay_h=00, sample_delay_l=02
Fx3UsbUsbCoreIsr
  PROT
    STATUS_STAGE
```
But when booted via USB:
```
req: c0 b2 value: 0000 index: 0000 length: 0001
CMD_GET_REVID_VERSION
Fx3UsbUsbCoreIsr
  PROT
    STATUS_STAGE
power val 80000001
power val 80000001
power val 80000001
power val 80000001
power val 80000001
power val 80000001
power val 80000001
power val 80000001
power val 80000001
power val 80000001
Fx3UsbUsbCoreIsr
  PROT
    SUTOK_EV
req: 40 b1 value: 0000 index: 0000 length: 0003
CMD_START
flags=80, sample_delay_h=00, sample_delay_l=03
Fx3UsbUsbCoreIsr
  PROT
    STATUS_STAGE
```
this looks like a silicon bug to me but :shrug: 

Would you be okay with removing the check inside while?

**Comment by zeldin** [comment]
So basically when running from RAM then the block is initially out of reset and active, which makes sense.  Forcing it into reset makes it not active, and then when you let it out of reset it's active again.  So far everything seems sane.

Now when you're running from SPI the block starts out in reset and not active.  This also seems reasonable.  The only weird thing is that when you release the reset, it does not become active.  Then when you make a new reset again it does not become active.  So this still feels like there is something that needs to be initialized first, which is implicitly done during USB boot but not SPI boot.

A relevant question is of course what the exact consequences of UIB not becoming "active" is, considering that you suggest that everything works seemingly correct even without it...  Did you try USB2 (High Speed) mode as well?  (IIRC "SUTOK_EV" means you are in the SuperSpeed path.)

**Comment by danielkucera** [comment]
Let me correct myself - when I don't deassert reset after SPI boot, it doesn't work (boot). After USB boot it works because it is runinng already.
The register behaves the same in USB-HS.
Additionaly, I found out it doesn't work (pulse view doesn't capture any samples) in USB-HS in any configuration.

**Comment by danielkucera** [comment]
you were right: https://community.infineon.com/t5/USB-superspeed-peripherals/FX3-UIB-power-register/m-p/688372#M36240

**Comment by danielkucera** [comment]
Hi @zeldin ,

what is the actual reason you are doing the power reset? Is it recommended in some startup procedure documentation or somewhere?

https://community.infineon.com/t5/USB-superspeed-peripherals/UIB-POWER-doesn-t-go-ACTIVE/m-p/1011340#M39799

**Comment by zeldin** [comment]
No, it's just what the SDK does in the `CyFx3BootUsbStart()` function, which is what my `Fx3UsbInit()` is attempting to replace.  I have not seen it mentioned in the HW docs, but they are rather incomplete when it comes to the USB stuff...

I do note that `CyFx3BootUsbStart()` does none of this if you pass `noReEnum` as TRUE, and USB is already active, but that shoudn't be your scenario since you say that USB is _not_ active when booting from SPI...

**Comment by danielkucera** [comment]
I was trying to find `CY_U3P_UIB_POWER` in the fx3 SDK but did not find it in the init. Could you please point me to the source code?

**Comment by zeldin** [comment]
The SDK does not provide source for the USB functions.  You'll need to disassemble (or decompile, if you have Ghidra :smile:) if you want to read it.

The function that contains this code is not large though, it's the function `CyFx3UsbPowerOn` in the object file `fx3_usb.o` in `libcyfxapi.a`.  `UIB_POWER` is accessed using the addressing mode `[r4, #3844]` with `r4` containing `0xe0017000` and `3844` being `0xf04`, thus forming the complete address `0xe0017f04`.

Hm, wait a minute...  That's the address of `PIB_POWER`, not `UIB_POWER`.  It looks like I maybe had those two registers mixed up all along?!

**Comment by zeldin** [comment]
Hm, no in my disassembly from before `r4` does contain `0xe0037000`.  I need to double check if there are multiple versions of the same function; maybe they fixed a bug and forgot to update all copies of the lib in the SDK?

**Comment by danielkucera** [comment]
interesting...

**Comment by danielkucera** [comment]
Mine shows this:
![Image](https://github.com/user-attachments/assets/60630665-f0b6-4fa0-a3fa-7dcb3cda9008)

this is taken from https://github.com/Zerithious/fx3-1.3.3-sdk
both
```
./sdk/firmware/src/fx3lib/fx3_release/cyfx3lib.a
./sdk/firmware/src/fx3lib/fx3_debug/cyfx3lib.a
```

**Comment by zeldin** [comment]
Yeah, that clearly has e0037f04 (`UIB_POWER`).  And now I think I was just confused; the disassembly I looked at now has both 0xe0037000 and 0xe0031000 as constants at the end of the function, and both are loaded into `r4` and used as base address at different points in the code, so I probably just mixed up (_this time_) which code was using which constant.

I should still run a check to see if all copies of the function are identical, because now I don't trust my own eyes anymore.  :grin:

**Comment by danielkucera** [comment]
Do you think it may be skipped due to false condition with the original sdk?

https://github.com/Zerithious/fx3-1.3.3-sdk/blob/b35a28237d69e563b89af42adad182804993b7b2/sdk/firmware/src/usb/cyu3usb.c#L658-L663

**Comment by zeldin** [comment]
Not so likely perhaps; I mean that check only makes a difference if you're using an OTG port in DFP ("host") mode, which seems like the less common use case for this chip, right?

**Comment by 0rt** [comment]
I am sorry to interrupt. If I just have a board of fx3, do I have to build my own pulseview as well as the fx3 firmware to make it works as a logic analyzer?

**Comment by zeldin** [comment]
@0rt I don't know why you are hijacking this issue instead of creating a new one, but yes, you need a custom `libsigrok` (which is what pulseview uses to access the hardware) with FX3 support, yes.  It is not in the upstream sources.

**Comment by 0rt** [comment]
sorry about that. because my problem is prefectly fit this issue tittle... but mostly because I had downloaded a firmware from @danielkucera , but my pulseview don't' load it.  So by any chance you guys are so kind that can distribute a prebuilt binary of working as well, for it looks like not a easy job to build them from sources.

**Comment by zeldin** [comment]
> So by any chance you guys are so kind that can distribute a prebuilt binary of working as well, for it looks like not a easy job to build them from sources.

Sure, do you want a PPC64 or a RISC-V binary?

**Comment by danielkucera** [comment]
Hi @zeldin ,

#6 confirms it is not just my HW specific case. 

I also tried to pursue the debugging with Cypress support but it was a dead end since this is not using their SDK.

Would you consider making the UIB reset "optional" or with some timeout in the loop?

**Comment by zeldin** [comment]
@danielkucera I just pushed code to implement a timeout, please check if it works for you.

**Comment by danielkucera** [comment]
Thanks @zeldin , https://github.com/zeldin/fx3lafw/commit/d65aaa7cb0d0aeddf0be41162dd9c3cee5190e04 works fine for me.
